uniffi_bindgen/src/bindings/gecko_js/templates/macros.cpp

```diff
@@ -35,25 +59,37 @@
-  if ({{ err }}.mCode) {
-    {%- match func.cpp_throw_by() %}
-    {%- when ThrowBy::ErrorResult with (rv) %}
-    {# /* TODO: Improve error throwing. See https://github.com/mozilla/uniffi-rs/issues/295
-          for details. */ -#}
-    {{ rv }}.ThrowOperationError(nsDependentCString({{ err }}.mMessage));
-    {%- when ThrowBy::Assert %}
-    MOZ_ASSERT(false);
-    {%- endmatch %}
-    return {% match func.cpp_return_type() %}{% when Some with (type_) %}{{ type_|dummy_ret_value_cpp(context) }}{% else %}{% endmatch %};
-  }
-  {%- match func.cpp_return_by() %}
-  {%- when ReturnBy::OutParam with (name, type_) %}
-  DebugOnly<bool> ok_ = {{ type_|lift_cpp(result, name, context) }};
-  MOZ_ASSERT(ok_);
-  {%- when ReturnBy::Value with (type_) %}
-  {{ type_|type_cpp(context) }} retVal_;
-  DebugOnly<bool> ok_ = {{ type_|lift_cpp(result, "retVal_", context) }};
-  MOZ_ASSERT(ok_);
-  return retVal_;
-  {%- when ReturnBy::Void %}{%- endmatch %}
-{%- endmacro -%}
+        if ({{ err }}.mCode) {
+          {%- match func.cpp_throw_by() %}
+          {%- when ThrowBy::ErrorResult with (rv) %}
+          {# /* TODO: Improve error throwing. See https://github.com/mozilla/uniffi-rs/issues/295 for details. XXXdmose code below is different now that we're async */ -#}
+          return MozPromise<{% match func.ffi_func().return_type() %} {% when Some with (type_) %}{{ type_|type_ffi(context) }} {% else %} int8_t{% endmatch %}, {{ context.ffi_rusterror_type() }}, false>::CreateAndReject(std::move({{ err }}), __func__);
+          {%- when ThrowBy::Assert %}
+          {%- endmatch %}
+        }
+
+        return MozPromise<{% match func.ffi_func().return_type() %}{% when Some with (type_) %}{{ type_|type_ffi(context) }} {% else %}int8_t{% endmatch %}, {{ context.ffi_rusterror_type() }}, false>::CreateAndResolve(std::move({{ result }}), __func__);
+  })->Then(GetCurrentSerialEventTarget(), __func__,
+    [promise]({% match func.ffi_func().return_type() %}{% when Some with (type_) %}const {{ type_|type_ffi(context) }} {% else %}const int8_t{% endmatch %} {{result}}) {
+      /* resolve DOM promise */
+
+      {%- match func.cpp_return_by() %}
+      {%- when ReturnBy::OutParam with (name, type_) %}
+      DebugOnly<bool> ok_ = {{ type_|lift_cpp(result, name, context) }};
+      MOZ_ASSERT(ok_);
+      {%- when ReturnBy::Value with (type_) %}
+      {{ type_|type_cpp(context) }} retVal_;
+      DebugOnly<bool> ok_ = {{ type_|lift_cpp(result, "retVal_", context) }};
+      MOZ_ASSERT(ok_);
+      promise->MaybeResolve(retVal_);
+      {%- when ReturnBy::Void %}
+      promise->MaybeResolveWithUndefined();
+      {%- endmatch %}
+    },
+    [promise]({{ context.ffi_rusterror_type() }} err) {
+      // XXX put the message into the error
+      // (a la aRv.ThrowOperationError(nsDependentCString(err.mMessage))
+      promise->MaybeReject(NS_ERROR_FAILURE);
+    });
+  return promise.forget(); // XXX i assume .forget is needed, check on this
```

Comment:
Pretty sure this is correct - although I'm a very very long way from any kind of expert here, IIUC, `Promise::Create()` returns an `already_AddRefed<>` and you don't need any other references, so .`forget()` is the right thing.